File: javascript/packages/highlighter/test/diff-renderer.test.ts
import { describe, it, expect, beforeEach } from "vitest"
import dedent from "dedent"

import { Herb } from "@herb-tools/node-wasm"
import { themes } from "../src/themes.js"
import { ANSI_REGEX } from "../src/ansi.js"
import { stripAnsiColors } from "./util.js"

import { DiffRenderer } from "../src/diff-renderer.js"
import { computeDiffHunks } from "../src/diff-computer.js"
import { SyntaxRenderer } from "../src/syntax-renderer.js"

const hexBackground = (hex: string) =>
  `\x1b[48;2;${parseInt(hex.slice(1, 3), 16)};${parseInt(hex.slice(3, 5), 16)};${parseInt(hex.slice(5, 7), 16)}m`

const ADDED_SPAN = hexBackground(themes.onedark.DIFF_ADDED_BACKGROUND as string)
const REMOVED_SPAN = hexBackground(themes.onedark.DIFF_REMOVED_BACKGROUND as string)

describe("DiffRenderer", () => {
  let renderer: DiffRenderer

  beforeEach(async () => {
    const syntaxRenderer = new SyntaxRenderer(themes.onedark, Herb)
    await syntaxRenderer.initialize()
    renderer = new DiffRenderer(syntaxRenderer, themes.onedark)
  })

  const original = dedent`
    <div id="gems">
      <span class='card'>
        <%= render partial: "gem_card" %>
      </span>
      <img src="a.png">
    </div>
  `

  const modified = dedent`
    <div id="gems">
      <span class="card">
        <%= render partial: "gem_card" %>
      </span>
      <img src="a.png" alt="">
    </div>
  `

  it("renders nothing when the sources are identical", () => {
    expect(renderer.render("/test/file.erb", original, original)).toBe("")
  })

  it("renders removals and additions in the gutter alongside line numbers", () => {
    const result = renderer.render("/test/file.erb", original, modified, { wrapLines: false, contextLines: 1 })

    expect(stripAnsiColors(result)).toMatchSnapshot()
  })

  it("keeps context numbering monotonic when the fix changes the line count", () => {
    const shorter = dedent`
      <div>
        <span>one</span>
      </div>
    `

    const longer = dedent`
      <div>
        <span>one</span>
        <span>two</span>
      </div>
    `

    const result = renderer.render("/test/file.erb", shorter, longer, { wrapLines: false })

    expect(stripAnsiColors(result)).toMatchSnapshot()
  })

  it("separates distant hunks with a marker", () => {
    const before = ["<div>", "<a>", "<b>", "<c>", "<d>", "<e>", "<f>"].join("\n")
    const after = ["<div>", "<a2>", "<b>", "<c>", "<d>", "<e2>", "<f>"].join("\n")

    const result = renderer.render("/test/file.erb", before, after, { wrapLines: false, contextLines: 1 })

    expect(stripAnsiColors(result)).toContain("⋮")
    expect(stripAnsiColors(result)).toMatchSnapshot()
  })

  it("omits the header and gutter when line numbers are hidden", () => {
    const result = renderer.render("/test/file.erb", original, modified, {
      wrapLines: false,
      contextLines: 1,
      showLineNumbers: false,
    })

    expect(stripAnsiColors(result)).toMatchSnapshot()
  })

  it("indents the whole diff for nesting under an offense", () => {
    const result = renderer.render("", original, modified, {
      wrapLines: false,
      contextLines: 0,
      indent: "      ",
    })

    expect(stripAnsiColors(result)).toMatchSnapshot()
  })

  it("truncates long lines when asked", () => {
    const before = `<div class="${"a".repeat(200)}">old</div>`
    const after = `<div class="${"a".repeat(200)}">new</div>`

    const result = renderer.render("/test/file.erb", before, after, { truncateLines: true, maxWidth: 60 })

    expect(stripAnsiColors(result)).toContain("…")
    expect(stripAnsiColors(result)).toMatchSnapshot()
  })

  describe("inline change highlighting", () => {
    it("backgrounds only the characters that changed, keeping syntax colors", () => {
      const result = renderer.render("/test/file.erb", original, modified, { wrapLines: false, contextLines: 0 })

      expect(result).toMatch(ANSI_REGEX)
      expect(result).toMatchSnapshot()
    })

    it("leaves lines unmarked when a block was restructured rather than rewritten in place", () => {
      const before = dedent`
        <div>
          <span>content</span>
        </div>
      `

      const after = dedent`
        <div>
          <section>
            <span>content</span>
          </section>
        </div>
      `

      const result = renderer.render("/test/file.erb", before, after, { wrapLines: false, contextLines: 0 })

      expect(result).not.toContain(REMOVED_SPAN)
      expect(result).not.toContain(ADDED_SPAN)
    })

    it("marks the inserted indentation when a block is re-indented line for line", () => {
      const before = ["<div>", "<span>one</span>", "<span>two</span>", "</div>"].join("\n")
      const after = ["<div>", "  <span>one</span>", "  <span>two</span>", "</div>"].join("\n")

      const result = renderer.render("/test/file.erb", before, after, { wrapLines: false, contextLines: 0 })

      expect(result).toContain(ADDED_SPAN)
      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("can be turned off", () => {
      const result = renderer.render("/test/file.erb", original, modified, {
        wrapLines: false,
        contextLines: 0,
        highlightInlineChanges: false,
      })

      expect(result).not.toContain(REMOVED_SPAN)
      expect(result).not.toContain(ADDED_SPAN)
    })
  })

  describe("split layout", () => {
    it("puts the original on the left and the modified on the right", () => {
      const result = renderer.render("", original, modified, { layout: "split", contextLines: 1, maxWidth: 140 })

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("numbers each column from its own version of the file", () => {
      const before = ["<div>", "  <span>x</span>", "</div>"].join("\n")
      const after = ["<div>", "  <section>", "    <span>x</span>", "  </section>", "</div>"].join("\n")

      const result = stripAnsiColors(renderer.render("", before, after, { layout: "split", contextLines: 1, maxWidth: 140 }))

      const lastRow = result.split("\n").at(-1)!
      const [left, right] = lastRow.split("┃")

      expect(left).toContain("3 │ </div>")
      expect(right).toContain("5 │ </div>")
    })

    it("leaves the shorter side blank when the two runs differ in length", () => {
      const before = ["<div>", "  <span>x</span>", "</div>"].join("\n")
      const after = ["<div>", "  <section>", "    <span>x</span>", "  </section>", "</div>"].join("\n")

      const result = stripAnsiColors(renderer.render("", before, after, { layout: "split", contextLines: 0, maxWidth: 140 }))

      expect(result).toMatchSnapshot()
    })

    it("falls back to the unified layout when the columns would be too narrow", () => {
      const result = stripAnsiColors(renderer.render("", original, modified, { layout: "split", contextLines: 1, maxWidth: 70 }))

      expect(result).not.toContain("┃")
      expect(result).toContain("-   2")
    })
  })

  describe("single-line collapsing", () => {
    const isolated = {
      before: dedent`
        <div>
          <span class='a'>one</span>
          <p>untouched</p>
        </div>
      `,
      after: dedent`
        <div>
          <span class="a">one</span>
          <p>untouched</p>
        </div>
      `,
    }

    it("collapses an isolated one-for-one replacement onto a single line", () => {
      const result = renderer.render("", isolated.before, isolated.after, {
        wrapLines: false,
        contextLines: 1,
        singleLineStyle: "inline",
      })

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("splits the same change by default", () => {
      const result = renderer.render("", isolated.before, isolated.after, { wrapLines: false, contextLines: 1 })

      expect(stripAnsiColors(result)).toContain("-   2")
      expect(stripAnsiColors(result)).toContain("+")
    })

    it("declines to collapse when a line has two separate edits", () => {
      const before = ["<div>", "  <span class='a' id='b'>x</span>", "</div>"].join("\n")
      const after = ["<div>", `  <span class="a" id="b">x</span>`, "</div>"].join("\n")

      const result = renderer.render("", before, after, { wrapLines: false, contextLines: 1, singleLineStyle: "inline" })

      expect(stripAnsiColors(result)).not.toContain("±")
    })

    it("declines to collapse a multi-line restructure", () => {
      const before = ["<div>", "  <span>x</span>", "</div>"].join("\n")
      const after = ["<div>", "  <section>", "    <span>x</span>", "  </section>", "</div>"].join("\n")

      const result = renderer.render("", before, after, { wrapLines: false, contextLines: 1, singleLineStyle: "inline" })

      expect(stripAnsiColors(result)).not.toContain("±")
    })

    it("falls back to splitting when color is off, since the composite needs tinting to be read", () => {
      const originalNoColor = process.env.NO_COLOR
      process.env.NO_COLOR = "1"

      try {
        const result = renderer.render("", isolated.before, isolated.after, {
          wrapLines: false,
          contextLines: 1,
          singleLineStyle: "inline",
        })

        expect(result).not.toContain("±")
        expect(result).not.toMatch(ANSI_REGEX)
        expect(result).toMatchSnapshot()
      } finally {
        if (originalNoColor === undefined) {
          delete process.env.NO_COLOR
        } else {
          process.env.NO_COLOR = originalNoColor
        }
      }
    })
  })

  describe("auto collapsing", () => {
    const wrap = (line: string) => ["<div>", line, "</div>"].join("\n")

    const collapses = (before: string, after: string, maxWidth = 120) =>
      renderer.render("", wrap(before), wrap(after), {
        contextLines: 0,
        wrapLines: false,
        singleLineStyle: "auto",
        maxWidth,
      }).includes("±")

    it("collapses a pure insertion, whose composite is a real line", () => {
      expect(collapses(`  <img src="a.png">`, `  <img src="a.png" alt="">`)).toBe(true)
    })

    it("collapses a pure deletion", () => {
      expect(collapses(`  <img src="a.png" onclick="x()">`, `  <img src="a.png">`)).toBe(true)
    })

    it("collapses a small replacement", () => {
      expect(collapses(`  <span class='a'>one</span>`, `  <span class="a">one</span>`)).toBe(true)
    })

    it("keeps a long replacement split", () => {
      expect(collapses(
        `  <span class='alpha beta gamma delta'>one</span>`,
        `  <span class="totally different set of names here">one</span>`,
      )).toBe(false)
    })

    it("keeps a replacement covering most of the line split", () => {
      expect(collapses(`  <a href='/old'>Old</a>`, `  <button data-x='y'>New</button>`)).toBe(false)
    })

    it("keeps a change split when the composite would not fit the width", () => {
      const before = `  <span class='a'>${"x".repeat(50)}</span>`
      const after = `  <span class="a">${"x".repeat(50)}</span>`

      expect(collapses(before, after, 120)).toBe(true)
      expect(collapses(before, after, 80)).toBe(false)
    })

    it("collapses nothing when color is off", () => {
      const originalNoColor = process.env.NO_COLOR
      process.env.NO_COLOR = "1"

      try {
        expect(collapses(`  <img src="a.png">`, `  <img src="a.png" alt="">`)).toBe(false)
      } finally {
        if (originalNoColor === undefined) {
          delete process.env.NO_COLOR
        } else {
          process.env.NO_COLOR = originalNoColor
        }
      }
    })
  })

  describe("renderFromHunks", () => {
    it("renders hunks that arrived without their sources", () => {
      const hunks = computeDiffHunks(original, modified, 1)
      const result = renderer.renderFromHunks("/test/file.erb", hunks, { wrapLines: false })

      expect(stripAnsiColors(result)).toMatchSnapshot()
    })

    it("matches what rendering from the sources produces", () => {
      const hunks = computeDiffHunks(original, modified, 1)

      const fromHunks = renderer.renderFromHunks("/test/file.erb", hunks, { wrapLines: false, contextLines: 1 })
      const fromSources = renderer.render("/test/file.erb", original, modified, { wrapLines: false, contextLines: 1 })

      expect(stripAnsiColors(fromHunks)).toBe(stripAnsiColors(fromSources))
    })

    it("renders nothing when given no hunks", () => {
      expect(renderer.renderFromHunks("/test/file.erb", [])).toBe("")
    })
  })

  describe("NO_COLOR environment", () => {
    it("renders the diff without any escape codes", () => {
      const originalNoColor = process.env.NO_COLOR
      process.env.NO_COLOR = "1"

      try {
        const result = renderer.render("/test/file.erb", original, modified, { wrapLines: false, contextLines: 1 })

        expect(result).not.toMatch(ANSI_REGEX)
        expect(result).toMatchSnapshot()
      } finally {
        if (originalNoColor === undefined) {
          delete process.env.NO_COLOR
        } else {
          process.env.NO_COLOR = originalNoColor
        }
      }
    })
  })
})

File: rust/herb-highlighter/tests/diff_renderer_test.rs
mod common;

use herb_highlighter::ansi::find_ansi_sequences;
use herb_highlighter::diff_computer::compute_diff_hunks;
use herb_highlighter::diff_renderer::{DiffLayout, DiffRenderOptions, DiffRenderer, SingleLineStyle};
use herb_highlighter::syntax_renderer::SyntaxRenderer;
use herb_highlighter::themes::{get_theme, Theme};

use common::{no_color, strip_ansi_colors, with_color};

const ORIGINAL: &str = "<div id=\"gems\">\n  <span class='card'>\n    <%= render partial: \"gem_card\" %>\n  </span>\n  <img src=\"a.png\">\n</div>";

const MODIFIED: &str = "<div id=\"gems\">\n  <span class=\"card\">\n    <%= render partial: \"gem_card\" %>\n  </span>\n  <img src=\"a.png\" alt=\"\">\n</div>";

fn syntax_renderer() -> SyntaxRenderer {
  SyntaxRenderer::new(get_theme(Theme::OneDark).clone())
}

fn added_span() -> String {
  get_theme(Theme::OneDark)
    .diff_added_background
    .expect("onedark carries diff colors")
    .background()
    .into_owned()
}

fn removed_span() -> String {
  get_theme(Theme::OneDark)
    .diff_removed_background
    .expect("onedark carries diff colors")
    .background()
    .into_owned()
}

#[test]
fn renders_nothing_when_the_sources_are_identical() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  assert_eq!(renderer.render("/test/file.erb", ORIGINAL, ORIGINAL, &DiffRenderOptions::default()), "");
}

#[test]
fn renders_removals_and_additions_in_the_gutter_alongside_line_numbers() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "/test/file.erb",
    ORIGINAL,
    MODIFIED,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn keeps_context_numbering_monotonic_when_the_fix_changes_the_line_count() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let shorter = "<div>\n  <span>one</span>\n</div>";
  let longer = "<div>\n  <span>one</span>\n  <span>two</span>\n</div>";

  let result = renderer.render(
    "/test/file.erb",
    shorter,
    longer,
    &DiffRenderOptions {
      wrap_lines: false,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn separates_distant_hunks_with_a_marker() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let before = ["<div>", "<a>", "<b>", "<c>", "<d>", "<e>", "<f>"].join("\n");
  let after = ["<div>", "<a2>", "<b>", "<c>", "<d>", "<e2>", "<f>"].join("\n");

  let result = renderer.render(
    "/test/file.erb",
    &before,
    &after,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      ..Default::default()
    },
  );

  assert!(strip_ansi_colors(&result).contains('⋮'));

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn omits_the_header_and_gutter_when_line_numbers_are_hidden() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "/test/file.erb",
    ORIGINAL,
    MODIFIED,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      show_line_numbers: false,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn indents_the_whole_diff_for_nesting_under_an_offense() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "",
    ORIGINAL,
    MODIFIED,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 0,
      indent: "      ".to_string(),
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn truncates_long_lines_when_asked() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let before = format!("<div class=\"{}\">old</div>", "a".repeat(200));
  let after = format!("<div class=\"{}\">new</div>", "a".repeat(200));

  let result = renderer.render(
    "/test/file.erb",
    &before,
    &after,
    &DiffRenderOptions {
      truncate_lines: true,
      max_width: Some(60),
      ..Default::default()
    },
  );

  assert!(strip_ansi_colors(&result).contains('…'));

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn backgrounds_only_the_characters_that_changed_keeping_syntax_colors() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "/test/file.erb",
    ORIGINAL,
    MODIFIED,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 0,
      ..Default::default()
    },
  );

  assert!(!find_ansi_sequences(&result).is_empty());

  insta::assert_snapshot!(result);
}

#[test]
fn leaves_lines_unmarked_when_a_block_was_restructured_rather_than_rewritten_in_place() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let before = "<div>\n  <span>content</span>\n</div>";
  let after = "<div>\n  <section>\n    <span>content</span>\n  </section>\n</div>";

  let result = renderer.render(
    "/test/file.erb",
    before,
    after,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 0,
      ..Default::default()
    },
  );

  assert!(!result.contains(&removed_span()));
  assert!(!result.contains(&added_span()));
}

#[test]
fn marks_the_inserted_indentation_when_a_block_is_reindented_line_for_line() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let before = ["<div>", "<span>one</span>", "<span>two</span>", "</div>"].join("\n");
  let after = ["<div>", "  <span>one</span>", "  <span>two</span>", "</div>"].join("\n");

  let result = renderer.render(
    "/test/file.erb",
    &before,
    &after,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 0,
      ..Default::default()
    },
  );

  assert!(result.contains(&added_span()));

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn inline_change_highlighting_can_be_turned_off() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "/test/file.erb",
    ORIGINAL,
    MODIFIED,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 0,
      highlight_inline_changes: false,
      ..Default::default()
    },
  );

  assert!(!result.contains(&removed_span()));
  assert!(!result.contains(&added_span()));
}

#[test]
fn puts_the_original_on_the_left_and_the_modified_on_the_right() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "",
    ORIGINAL,
    MODIFIED,
    &DiffRenderOptions {
      layout: DiffLayout::Split,
      context_lines: 1,
      max_width: Some(140),
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn numbers_each_column_from_its_own_version_of_the_file() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let before = ["<div>", "  <span>x</span>", "</div>"].join("\n");
  let after = ["<div>", "  <section>", "    <span>x</span>", "  </section>", "</div>"].join("\n");

  let result = strip_ansi_colors(&renderer.render(
    "",
    &before,
    &after,
    &DiffRenderOptions {
      layout: DiffLayout::Split,
      context_lines: 1,
      max_width: Some(140),
      ..Default::default()
    },
  ));

  let last_row = result.split('\n').next_back().expect("the diff has rows").to_string();
  let mut columns = last_row.split('┃');

  assert!(columns.next().expect("a left column").contains("3 │ </div>"));
  assert!(columns.next().expect("a right column").contains("5 │ </div>"));
}

#[test]
fn leaves_the_shorter_side_blank_when_the_two_runs_differ_in_length() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let before = ["<div>", "  <span>x</span>", "</div>"].join("\n");
  let after = ["<div>", "  <section>", "    <span>x</span>", "  </section>", "</div>"].join("\n");

  let result = strip_ansi_colors(&renderer.render(
    "",
    &before,
    &after,
    &DiffRenderOptions {
      layout: DiffLayout::Split,
      context_lines: 0,
      max_width: Some(140),
      ..Default::default()
    },
  ));

  insta::assert_snapshot!(result);
}

#[test]
fn falls_back_to_the_unified_layout_when_the_columns_would_be_too_narrow() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = strip_ansi_colors(&renderer.render(
    "",
    ORIGINAL,
    MODIFIED,
    &DiffRenderOptions {
      layout: DiffLayout::Split,
      context_lines: 1,
      max_width: Some(70),
      ..Default::default()
    },
  ));

  assert!(!result.contains('┃'));
  assert!(result.contains("-   2"));
}

const ISOLATED_BEFORE: &str = "<div>\n  <span class='a'>one</span>\n  <p>untouched</p>\n</div>";
const ISOLATED_AFTER: &str = "<div>\n  <span class=\"a\">one</span>\n  <p>untouched</p>\n</div>";

#[test]
fn collapses_an_isolated_one_for_one_replacement_onto_a_single_line() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "",
    ISOLATED_BEFORE,
    ISOLATED_AFTER,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      single_line_style: SingleLineStyle::Inline,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn splits_the_same_change_by_default() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = strip_ansi_colors(&renderer.render(
    "",
    ISOLATED_BEFORE,
    ISOLATED_AFTER,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      ..Default::default()
    },
  ));

  assert!(result.contains("-   2"));
  assert!(result.contains('+'));
}

#[test]
fn declines_to_collapse_when_a_line_has_two_separate_edits() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let before = ["<div>", "  <span class='a' id='b'>x</span>", "</div>"].join("\n");
  let after = ["<div>", "  <span class=\"a\" id=\"b\">x</span>", "</div>"].join("\n");

  let result = strip_ansi_colors(&renderer.render(
    "",
    &before,
    &after,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      single_line_style: SingleLineStyle::Inline,
      ..Default::default()
    },
  ));

  assert!(!result.contains('±'));
}

#[test]
fn declines_to_collapse_a_multi_line_restructure() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let before = ["<div>", "  <span>x</span>", "</div>"].join("\n");
  let after = ["<div>", "  <section>", "    <span>x</span>", "  </section>", "</div>"].join("\n");

  let result = strip_ansi_colors(&renderer.render(
    "",
    &before,
    &after,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      single_line_style: SingleLineStyle::Inline,
      ..Default::default()
    },
  ));

  assert!(!result.contains('±'));
}

#[test]
fn falls_back_to_splitting_when_color_is_off() {
  no_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "",
    ISOLATED_BEFORE,
    ISOLATED_AFTER,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      single_line_style: SingleLineStyle::Inline,
      ..Default::default()
    },
  );

  assert!(!result.contains('±'));
  assert!(find_ansi_sequences(&result).is_empty());

  insta::assert_snapshot!(result);
}

fn collapses(before: &str, after: &str, max_width: usize) -> bool {
  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let wrap = |line: &str| ["<div>", line, "</div>"].join("\n");

  renderer
    .render(
      "",
      &wrap(before),
      &wrap(after),
      &DiffRenderOptions {
        context_lines: 0,
        wrap_lines: false,
        single_line_style: SingleLineStyle::Auto,
        max_width: Some(max_width),
        ..Default::default()
      },
    )
    .contains('±')
}

#[test]
fn auto_collapses_a_pure_insertion() {
  with_color();

  assert!(collapses("  <img src=\"a.png\">", "  <img src=\"a.png\" alt=\"\">", 120));
}

#[test]
fn auto_collapses_a_pure_deletion() {
  with_color();

  assert!(collapses("  <img src=\"a.png\" onclick=\"x()\">", "  <img src=\"a.png\">", 120));
}

#[test]
fn auto_collapses_a_small_replacement() {
  with_color();

  assert!(collapses("  <span class='a'>one</span>", "  <span class=\"a\">one</span>", 120));
}

#[test]
fn auto_keeps_a_long_replacement_split() {
  with_color();

  assert!(!collapses(
    "  <span class='alpha beta gamma delta'>one</span>",
    "  <span class=\"totally different set of names here\">one</span>",
    120
  ));
}

#[test]
fn auto_keeps_a_replacement_covering_most_of_the_line_split() {
  with_color();

  assert!(!collapses("  <a href='/old'>Old</a>", "  <button data-x='y'>New</button>", 120));
}

#[test]
fn auto_keeps_a_change_split_when_the_composite_would_not_fit_the_width() {
  with_color();

  let before = format!("  <span class='a'>{}</span>", "x".repeat(50));
  let after = format!("  <span class=\"a\">{}</span>", "x".repeat(50));

  assert!(collapses(&before, &after, 120));
  assert!(!collapses(&before, &after, 80));
}

#[test]
fn auto_collapses_nothing_when_color_is_off() {
  no_color();

  assert!(!collapses("  <img src=\"a.png\">", "  <img src=\"a.png\" alt=\"\">", 120));
}

#[test]
fn renders_hunks_that_arrived_without_their_sources() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let hunks = compute_diff_hunks(ORIGINAL, MODIFIED, 1);

  let result = renderer.render_from_hunks(
    "/test/file.erb",
    &hunks,
    &DiffRenderOptions {
      wrap_lines: false,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(strip_ansi_colors(&result));
}

#[test]
fn matches_what_rendering_from_the_sources_produces() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let hunks = compute_diff_hunks(ORIGINAL, MODIFIED, 1);

  let options = DiffRenderOptions {
    wrap_lines: false,
    context_lines: 1,
    ..Default::default()
  };

  let from_hunks = renderer.render_from_hunks("/test/file.erb", &hunks, &options);
  let from_sources = renderer.render("/test/file.erb", ORIGINAL, MODIFIED, &options);

  assert_eq!(strip_ansi_colors(&from_hunks), strip_ansi_colors(&from_sources));
}

#[test]
fn renders_nothing_when_given_no_hunks() {
  with_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  assert_eq!(renderer.render_from_hunks("/test/file.erb", &[], &DiffRenderOptions::default()), "");
}

#[test]
fn renders_the_diff_without_any_escape_codes_when_color_is_off() {
  no_color();

  let syntax_renderer = syntax_renderer();
  let renderer = DiffRenderer::new(&syntax_renderer, get_theme(Theme::OneDark).clone());

  let result = renderer.render(
    "/test/file.erb",
    ORIGINAL,
    MODIFIED,
    &DiffRenderOptions {
      wrap_lines: false,
      context_lines: 1,
      ..Default::default()
    },
  );

  assert!(find_ansi_sequences(&result).is_empty());

  insta::assert_snapshot!(result);
}
